### ISSUE TITLE
[1.8.1] Puppet not found in Win Guest PATH

### ISSUE THREAD

**Issue by stefanaj** [mitchellh/vagrant#6778]
Host: Windows 10
Guest: Windows 7
Virtualbox 5.0.12
Puppet 3.8.4

With vagrant 1.8.1 there seems to be problems with recognizing puppet being in the PATH of a Windows 7 Guest. I tried reinstalling Vagrant 1.7.4 and it runs the puppet provisioner fine.

I provision using the following lines in my Vagrantfile:
```
vm.provision "puppet" do |puppet|
	puppet.manifest_file  = machine_name
	puppet.module_path = "modules"
	puppet.hiera_config_path = "hiera.yaml"
end
```
Directory structure:
- dev
  - manifests
    - machine_name
      - 01_site.pp
      - 02_build.pp
    - other_machine_name
      - 01_site.pp
      - 02_build.pp
    - hieradata
      - config.yaml
  - modules
  - Vagrantfile
  - hiera.yaml

`machine_name` is a directory.

Has anyone successfully provisioned a windows machine using puppet and Vagrant 1.8.1?

**Comment by vainikkaj** [comment]
Yes, works as expected on windows2012r2.

Following is a cleaned debug log snippet from running puppet in windows2008r2:

```
DEBUG winrmshell: powershell executing:
$ProgressPreference='SilentlyContinue';
where puppet
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
DEBUG winrmshell: Output: {:data=>[{:stderr=>""}, {:stderr=>"#< CLIXML\r\n"},
{:stderr=>"
<Objs Version=\"1.1.0.1\" xmlns=\"http://schemas.microsoft.com/powershell/2004/04\">
<S S=\"Error\">Where-Object : Cannot bind parameter 'FilterScript'. Cannot convert the \"puppet_x000D__x000A_</S>
<S S=\"Error\">\" value of type \"System.String\" to type \"System.Management.Automation.ScriptBlo_x000D__x000A_</S>
<S S=\"Error\">ck\"._x000D__x000A_</S>
<S S=\"Error\">At line:2 char:6_x000D__x000A_</S>
<S S=\"Error\">+ where &lt;&lt;&lt;&lt;  puppet_x000D__x000A_</S>
<S S=\"Error\">    + CategoryInfo          : InvalidArgument: (:) [Where-Object], ParameterBi _x000D__x000A_</S>
<S S=\"Error\">   ndingException_x000D__x000A_</S>
<S S=\"Error\">    + FullyQualifiedErrorId : CannotConvertArgumentNoMessage,Microsoft.PowerSh _x000D__x000A_</S>
<S S=\"Error\">   ell.Commands.WhereObjectCommand_x000D__x000A_</S>
<S S=\"Error\"> _x000D__x000A_</S></Objs>
"}], :exitcode=>1}
```
http://serverfault.com/questions/566722/power-shell-error-cannot-bind-parameter

It seems that change made in https://github.com/mitchellh/vagrant/pull/6259 does not work in Powershell 2.0.

@philipwigg: Could you verify if this actually is the case?

**Comment by philipwigg** [comment]
Yes, it seems like that's the case.

**Comment by philipwigg** [comment]
Actually it seems like ```which``` was correct and I was wrong to change it. The ```which``` commands gets translated into the following bit of Powershell on Windows machines...

https://github.com/mitchellh/vagrant/blob/master/plugins/communicators/winrm/command_filters/which.rb

The problem with Puppet not being found was only when using WMF 5.0.

I think maybe the ```where``` command actually doesn't do anything but returns True so the run continues and 'works' because Puppet is in the path anyway.

I think maybe the ```which``` is failing with WMF 5.0 because it returns some data in STDERR so this line is returning false...

https://github.com/mitchellh/vagrant/blob/master/plugins/communicators/winrm/communicator.rb#L165

So, it kind of fixed WMF 5.0 but broke Powershell 2.0. :-(

I will see if I can fix it properly.

**Comment by stefanaj** [comment]
@vainikkaj @philipwigg Thanks! looking forward to a fix.

**Comment by jmnav** [comment]
Let me add a _"me too"_ to this one.
Host: Debian _"wheezy"_
Guest: Windows 7 Enterprise SP1 x86
Virtualbox: 5.0.10

On Vagrant 1.7.4 it works as expected.  On Vagrant 1.8.1 it throws _"The `puppet` binary appears not to be in the PATH of the guest."_

**Comment by philipwigg** [comment]
Dup of issue https://github.com/mitchellh/vagrant/issues/6876, now fixed by PR https://github.com/mitchellh/vagrant/pull/6912

**Comment by chrisroberts** [comment]
Closing as this was fixed via #6912
